src/Streamly/Internal/Data/Ring/Foreign.hs

```diff
@@ -506,3 +509,48 @@ unsafeFoldRingNM count rh f z rb@Ring {..} =
         if ptr == rh || n == 0
             then return acc'
             else go (n - 1) acc' ptr
+
+data Tuple4' a b c d = Tuple4' !a !b !c !d deriving Show
+
+-- | @rollWindow n collector@ is a simple, incremental sliding window 
+-- fold that does not require all the intermediate elements in a computation.
+-- This maintains n elements in the window, when a new element comes it slides
+-- out the oldest element and the new element along with the old element are 
+-- supplied to the collector fold. 
+--
+-- The Maybe is for the case when initially the window is filling and 
+-- there is no old element.
+--
+{-# INLINE rollWindow #-}
+rollWindow :: forall m a b. (MonadIO m, Storable a) 
+    => Int -> Fold m (a, Maybe a) b -> Fold m a b
+rollWindow n (Fold step1 initial1 extract1)= Fold step initial extract
+
+    where
+
+    initial = do 
+        s <- initial1
+        (rb, rh) <- liftIO $ new n
+        return $
+            case s of
+                Fold.Partial x -> Fold.Partial $ Tuple4' rb rh (0 :: Int) x
+                Fold.Done b -> Fold.Done b
+
+    step (Tuple4' rb rh i x') a 
+        | i < n = do
+            rh1 <- liftIO $ unsafeInsert rb rh a
+            s <- step1 x' (a, Nothing)
+            return $
+                case s of
+                    Fold.Partial x -> Fold.Partial $ Tuple4' rb rh1 (i + 1) x
+                    Fold.Done b -> Fold.Done b
+        | otherwise = do
+            a' <- liftIO $ peek rh
```

Comment:
We can use `old` instead of `a'` for better cue about what it is. We usually do not use prime on variable names, to represent new generation of a variable we use x1, x2 etc instead.

Reply:
Also, see the coding guidelines here https://github.com/composewell/streamly/blob/master/CONTRIBUTING.md .